sql/hive/src/main/scala/org/apache/spark/sql/hive/SQLBuilder.scala

```diff
@@ -211,6 +215,73 @@ class SQLBuilder(logicalPlan: LogicalPlan, sqlContext: SQLContext) extends Loggi
     )
   }
 
+  private def groupingSetToSQL(
+      plan: Aggregate,
+      expand: Expand,
+      project: Project): String = {
+    require(plan.groupingExpressions.length > 1)
+
+    // The last column of Expand is always grouping ID
+    val gid = expand.output.last
+
+    // In cube/rollup/groupingsets, Analyzer creates new aliases for all group by expressions.
+    // Since conversion from attribute back SQL ignore expression IDs, the alias of attribute
+    // references are ignored in aliasMap
+    val aliasMap = AttributeMap(project.projectList.collect {
+      case a @ Alias(child, name) => (a.toAttribute, a)
+    })
+
+    val groupByAttributes = plan.groupingExpressions.dropRight(1).map(_.asInstanceOf[Attribute])
+    val groupByAttrMap = AttributeMap(groupByAttributes.zip(
+      project.projectList.drop(project.child.output.length).map(_.asInstanceOf[Alias].child)))
+    val groupingExprs = groupByAttrMap.values.toArray
+    val groupingSQL = groupingExprs.map(_.sql).mkString(", ")
+
+    val groupingSet = expand.projections.map(_.dropRight(1).filter {
```

Comment:
This will generate Option. A little change I made:
```
    val groupingSet = expand.projections.map { project =>
      project.dropRight(1).collect {
        case attr: Attribute if groupingAttrSet.contains(attr) => groupByAttrMap(attr)
      }
    }
```
Thank you! Let me know if you do not like this change.